### PR TITLE
Fix: refine the config sync logic

### PR DESCRIPTION
**Pull request by zzxwill**
- Refine the config sync logic

### Description of your changes

Fixes #

I have:

- [ ] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

**Comment by zzxwill** [comment]
image

**Review by Somefive — APPROVED** [review]

File: pkg/apiserver/rest/usecase/application.go [hunk 1]
```diff
@@ -706,14 +706,18 @@ func (c *applicationUsecaseImpl) Deploy(ctx context.Context, app *model.Applicat
...
-	targets, err := listTarget(ctx, c.ds, app.Project, nil)
+	env, err := c.envUsecase.GetEnv(ctx, workflow.EnvName)
 	if err != nil {
 		return nil, err
 	}
 	var clusterTargets []*model.ClusterTarget
-	for i, t := range targets {
-		if t.Cluster != nil {
-			clusterTargets = append(clusterTargets, targets[i].Cluster)
+	for _, t := range env.Targets {
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/application.go:714]: Circular queries are too slow, please query once.

Comment (zzxwill) [reply pkg/apiserver/rest/usecase/application.go:714]: 什么意思？

Comment (barnettZQG) [reply pkg/apiserver/rest/usecase/application.go:714]: 通过一次查询取回所有 Target。

Comment (zzxwill) [reply pkg/apiserver/rest/usecase/application.go:714]: 被删掉的那样？
image

File: pkg/apiserver/rest/usecase/config.go [hunk 2]
```diff
@@ -278,6 +281,26 @@ func SyncConfigs(ctx context.Context, k8sClient client.Client, project string, t
 			return err
 		}
 		// config sync application doesn't exist, create one
+		clusterTargets := convertClusterTargets(targets)
+		if len(clusterTargets) == 0 {
+			errMsg := "no policy (no targets found) to sync configs"
+			klog.InfoS(errMsg, "project", project)
+			return errors.New(errMsg)
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/config.go:288]: If no targets, should return nil?

File: pkg/apiserver/rest/usecase/config.go [hunk 3]
```diff
@@ -340,32 +369,107 @@ func SyncConfigs(ctx context.Context, k8sClient client.Client, project string, t
...
+	var s []interface{}
+	for _, v := range i {
+		s = append(s, v)
+	}
+	return s
+}
+
+// destroySyncConfigsApp will delete the application which is used to sync configs
+func destroySyncConfigsApp(ctx context.Context, k8sClient client.Client, project string) error {
+	name := fmt.Sprintf("config-sync-%s", project)
```
Comment (wonderflow) [inline pkg/apiserver/rest/usecase/config.go:467]: let's make a const variable for this?

File: pkg/apiserver/rest/usecase/application.go (hunk 1 above)
Comment (barnettZQG) [reply pkg/apiserver/rest/usecase/application.go]: 先确定要获取哪些 Targets信息，然后通过一次库查询查回数据。

**Review by wonderflow — APPROVED** [review]
generally LGTM

File: pkg/apiserver/rest/usecase/config.go [hunk 4]
```diff
@@ -321,6 +345,11 @@ func SyncConfigs(ctx context.Context, k8sClient client.Client, project string, t
 	}
 
 	mergedTarget := mergeTargets(currentTargets, targets)
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/config.go:347]: I can get through here temporarily, but incrementally increasing delivery targets is still a risky behavior.

File: pkg/apiserver/rest/usecase/config.go (hunk 3 above)
Comment (zzxwill) [reply pkg/apiserver/rest/usecase/config.go:465]: Done.

File: pkg/apiserver/rest/usecase/config.go (hunk 2 above)
Comment (zzxwill) [reply pkg/apiserver/rest/usecase/config.go:288]: Without targets, the application won't have any policies, the application will fail which means the sync of config will fail. Users have to redeploy the application to make it work.

Comment (barnettZQG) [reply pkg/apiserver/rest/usecase/config.go:289]: I mean without targets there is no need for configuration synchronization and not block deploy the app.

**Review by barnettZQG — APPROVED** [review]

**Pull request merged**